Support listing available video modes for a monitor

aleksijuvani opened the pull request:
Resolves #877.

- [x] Tested on all platforms changed
     - [x]  Windows
     - [x]  macOS
     - [x]  X11
     - [x]  Wayland
     - [x]  iOS
     - [ ]  Android
     - [ ]  Emscripten
- [x] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
- [x] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [x] Created an example program if it would help users understand this functionality
- [x] Updated [feature matrix](https://github.com/rust-windowing/winit/blob/master/FEATURES.md), if new features were added or implemented

Adds a new method `get_video_modes` to the `MonitorId` struct in the public API, and implements it on the Windows platform. I'm going to look into implementing this on macOS as well..

Osspial commented:
Hey, thanks for the PR!

FYI before you get much further into implementing this on macOS - the vast majority of work is currently being done on the `eventloop-2.0` branch, and we're expecting to merge that onto master Soon. I'd suggest reworking this PR against that branch. Fortunately, looking at these changes I wouldn't expect that to be too difficult to do.

aleksijuvani commented:
Thanks for the heads up. I'll take a look.

ZeGentzy commented on src/lib.rs:
```diff
@@ -540,3 +542,42 @@ impl Default for WindowAttributes {
+pub struct VideoMode {
+    dimensions: (u32, u32),
+    bit_depth: u16,
+    frequency: u16,
```
you mean refresh rate? Frequency is a pretty vague name.

aleksijuvani replied:
I'm open to bikeshedding this name, but I've commonly seen frequency used as a synonym for the display refresh rate. It is measured in hertz after all, which indicate a frequency..

ZeGentzy replied:
Yeah, but frequency could be frequency of anything. Sound is also measured in hertz. I'd be okay with something like `refresh_frequency` or something.

aleksijuvani replied:
I see. I don't think I've seen it called `refresh_frequency` before. `refresh_rate` is pretty self-explainatory and would work for me.

ZeGentzy commented on src/lib.rs:
```diff
@@ -540,3 +542,42 @@ impl Default for WindowAttributes {
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct VideoMode {
+    dimensions: (u32, u32),
```
Can this become `PhysicalSize` instead of `(u32, u32)`

aleksijuvani replied:
This needs to implement `Hash` which `PhysicalSize` does not (and can not, as it's essentially a tuple of `f64`). Hence, we're doing the conversion from `(u32, u32)` to `PhysicalSize` in the getter method.

Osspial replied:
`PhysicalSize` is stored as `u32`s and not `f64`s on `eventloop-2.0`, so this should be possible once rebased.

aleksijuvani replied:
Looks like this would depend on the changes made in #895 being merged first.

ZeGentzy commented on src/lib.rs:
```diff
@@ -540,3 +542,42 @@ impl Default for WindowAttributes {
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct VideoMode {
+    dimensions: (u32, u32),
+    bit_depth: u16,
```
Is it possible to expose the number of bits for each channel separately? At least distinguish the number of color bits from alpha bits?

aleksijuvani replied:
WinAPI does not expose this information as far as I can tell. SDL has some heuristics for this, but I think those were written with 16-bit and 8-bit modes in mind for older Windows systems. There's still some code for greyscale support in there.. Windows 8 and up no longer supports these modes. What do you need this information for?

ZeGentzy replied:
No, reason, I just noticed glfw exposed the separate bits. That's OK.

aleksijuvani commented:
I've reworked this against the `eventloop-2.0` branch and added support for macOS. There's a bit of a quirk there, as we can't currently obtain the refresh rate of the monitor without some patches upstream against the core-video-sys crate for example.

I'm also not a fan of allocating an additional `HashSet` on macOS instead of just re-using the iterator we get from calling `map` on the `Vec` returned from `CGDisplayMode::all_display_modes`, but I couldn't come up with a clean way to do this.

How do we proceed with implementing this for the rest of the platforms? As for the Android and iOS platforms I think this comes down to just returning the screen size, and additionally obtaining the refresh rate and bit depth if applicable. I haven't looked into what Emscripten is doing.

ZeGentzy commented on src/monitor.rs:
```diff
@@ -44,6 +47,51 @@ impl Iterator for AvailableMonitorsIter {
+/// - [`MonitorHandle::video_modes`][monitor_get].
+///
+/// [monitor_get]: ../monitor/struct.MonitorHandle.html#method.video_modes
+pub struct VideoModesIter<'a> {
```
`type VideoModesIter<'a> = HashSetIter<'a, VideoMode>;`? No need to define our own struct and everything, I think.

aleksijuvani replied:
I think we ought to newtype it to avoid a breaking change if this type changes in the future, as the underlying collection, and hence the iterator type is an implementation detail. That is essentially what we're doing here.

ZeGentzy replied:
Would an type alias cause breaking changes if the details get changed? I wouldn't think so, but I haven't put much thought into the matter.

aleksijuvani replied:
I think it could, because it essentially acts the same as if the alias and the aliased type were equivalent.

As an example, let's say there's a function in user code that takes a `std::collections::hash_set::Iter` parameter (for some reason) instead of a generic `Iterator` parameter. If the alias in winit were suddenly to point to a `std::collections::vec_deque::Iter` instead, this would become a type check error. This is probably not how you would usually write a function that takes an iterator parameter, but it illustrates the problem.

I think different iterators do also expose different methods on them, so if some methods were suddenly to disappear, or new ones to appear entirely, this would be more visible of a breaking change.

ZeGentzy commented:
> I've reworked this against the `eventloop-2.0` branch and added support for macOS. There's a bit of a quirk there, as we can't currently obtain the refresh rate of the monitor without some patches upstream against the core-video-sys crate for example.

Filled upstream already, yeah?

> I'm also not a fan of allocating an additional `HashSet` on macOS instead of just re-using the iterator we get from calling `map` on the `Vec` returned from `CGDisplayMode::all_display_modes`, but I couldn't come up with a clean way to do this.

I guess you could use a vec, then sort it at the end (or always keep it sorted) and dedup it, however that sounds like a pain to maintain.

aleksijuvani commented:
> > I've reworked this against the `eventloop-2.0` branch and added support for macOS. There's a bit of a quirk there, as we can't currently obtain the refresh rate of the monitor without some patches upstream against the core-video-sys crate for example.
> 
> Filled upstream already, yeah?

I'll look into filing a pull request against core-video-sys for this today.

aleksijuvani commented:
We're now reporting the display refresh rate correctly on macOS. I've filed LuoZijun/rust-core-video-sys#2 to expose the necessary functionality from the Core Video API.

aleksijuvani commented:
LuoZijun/rust-core-video-sys#2 has been merged, so all that remains is implementing this for the remaining platforms. I suppose I could take a look at either iOS or Linux today, so that we can ship this as soon as possible.

aleksijuvani commented:
I've finished the iOS implementation. I couldn't get the Emscripten or Android platforms to compile. It looks like their implementation in eventloop-2.0 hasn't been finished.

aleksijuvani commented:
I've finished implementing this on Wayland, and found a cleaner way to return iterators from the `video_modes` method. I've refactored all platforms accordingly. I'm going to look into implementing this on X11 next.

aleksijuvani commented:
X11 implementation is done. This is now implemented on all platforms except for Android and Emscripten that don't seem to be ported over to eventloop-2.0 yet. Should be ready for final review!

aleksijuvani commented on src/platform_impl/windows/monitor.rs:
```diff
@@ -1,21 +1,23 @@
 
 /// Win32 implementation of the main `MonitorHandle` object.
-#[derive(Debug, Clone)]
+#[derive(Clone)]
```
I would've normally used derivative here, but it's not used elsewhere in winit, so I didn't want to introduce an additional dependency.

ZeGentzy replied:
While @Osspial might have a different opinion, I think derivative is the better option for maintainability here.

Osspial replied:
I'm fine with using derivative here.

aleksijuvani replied:
I've changed this to use derivative.

aleksijuvani commented on src/platform_impl/ios/monitor.rs:
```diff
@@ -134,6 +129,27 @@ impl Inner {
+            let size: CGSize = unsafe { msg_send![mode, size] };
+            modes.insert(VideoMode {
+                dimensions: (size.width as u32, size.height as u32),
+                bit_depth: 32,
```
There's no way to query the bit depth on iOS as far as I can tell. I believe the bit depth is always 32 bits.

aleksijuvani commented on src/platform_impl/linux/wayland/event_loop.rs:
```diff
@@ -584,6 +585,20 @@ impl MonitorHandle {
+            .into_iter()
+            .map(|x| VideoMode {
+                dimensions: (x.dimensions.0 as u32, x.dimensions.1 as u32),
+                refresh_rate: (x.refresh_rate as f32 / 1000.0 + 0.5) as u16,
```
We're receiving the refresh rate from Wayland in millihertz, so we must divide it by a thousand to get it in hertz. We then effectively `ceil` the float before casting it into an integer, but I suppose we could've used `f32::ceil` here also.

vberger replied:
In all, I'm not sure it is the right approach to represent refresh rates as `u16` in Hertz. There are displays whose refresh rate is not an integer in this unit, no?

aleksijuvani replied:
The refresh rate is usually not an exact integer, but very close (e.g. 59.996 Hz). We could report the display refresh rate as a float, but Windows and iOS only report it as an integer, so it might be misleading to return a float on these platforms. I assume this is what you meant?

vberger replied:
I don't get what would be misleading about using a float while some platform always return integer values, could you elaborate on that?

We already have precedent with the DPI factor values, which are always integer on Wayland for example.

aleksijuvani replied:
To elaborate, I feel like that returning a float here implies that this is an exact refresh rate, when in fact it's truncated or rounded (on at least two platforms). Returning an integer type does not have this implication to me.

If we look at other windowing libraries, there's precedent of using integers to represent the refresh rate in SDL and GLFW. I don't feel particularly strongly about this either way, so I'm fine with changing this to return floats if we all agree on that.

vberger replied:
I personally have the other expectation, imo an integer refresh rate suggests that it is exact...

Anyway, I don't know much about which uses an app would have for the refresh rate. If rounding it doe snot pose any functionality issue, then I'm fine with it.

aleksijuvani replied:
Having slept on it, I think we should leave it as a rounded integer, so at least the behaviour is consistent across platforms. If the user requires an exact value, there's not much point in us returning an exact value on some platforms, and a rounded value on others. Furthermore, changing this to a floating point value would complicate the implementation (though slightly) for `VideoMode` as we would now have to implement `Hash` and `Eq` by hand. If there later appears an use-case where an exact value is required, we can revisit this. For now, perhaps we can document it in the function description that the returned value is an approximation?

aleksijuvani commented on src/platform_impl/linux/x11/util/randr.rs:
```diff
@@ -114,6 +115,33 @@ impl XConnection {
+            .filter(|x| output_modes.iter().any(|id| x.id == *id))
+            .map(|x| {
+                let refresh_rate = if x.dotClock > 0 && x.hTotal > 0 && x.vTotal > 0 {
+                    x.dotClock as u64 * 1000 / (x.hTotal as u64 * x.vTotal as u64)
```
I don't know what's going on here, but this is the same line of code that SDL, GTK and GLFW use to calculate the display refresh rate. I couldn't find any documentation for this.

aleksijuvani commented on src/platform_impl/linux/x11/util/randr.rs:
```diff
@@ -114,6 +115,33 @@ impl XConnection {
+
+                VideoMode {
+                    dimensions: (x.width, x.height),
+                    refresh_rate: (refresh_rate as f32 / 1000.0 + 0.5) as u16,
```
Same thing as in Wayland. We're converting from millihertz to hertz and ceiling the value.

aleksijuvani commented on src/platform_impl/macos/monitor.rs:
```diff
@@ -93,6 +102,44 @@ impl MonitorHandle {
+            .expect("failed to obtain list of display modes")
+            .into_iter()
+            .map(move |mode| {
+                let cg_refresh_rate = (mode.refresh_rate() + 0.5) as i64;
```
Yes, that should work. I'll see if I can change these to use `ceil` instead.

aleksijuvani replied:
Actually, this isn't a `ceil`, but `round`. Same goes for the other occurrences. I've changed the code to use `round`.

aloucks commented:
It makes it difficult to see what's going on when you force-push after comments have started. The commits will all get squashed in the end anyway.

aleksijuvani commented:
Sorry about that. I didn't realize this repository squashes pull requests upon merge. While making changes to the Wayland platform, I noticed that I needed to restructure the other platforms as well, and I decided to squash it all into a single commit at that point. There shouldn't be any more large changes landing to this branch at this point.

Osspial commented on FEATURES.md:
```diff
@@ -156,6 +156,7 @@ Legend:
 |Window maximization toggle       |✔️     |✔️     |✔️         |✔️             |**N/A**|**N/A**|**N/A**   |
 |Fullscreen                       |✔️     |✔️     |✔️         |✔️             |**N/A**|**N/A**|❌        |
 |Fullscreen toggle                |✔️     |✔️     |✔️         |✔️             |**N/A**|**N/A**|❌        |
+|Video mode query                 |✔️     |✔️     |✔️         |✔️             |❌    |✔️     |❌        |
```
Could you add a brief blurb describing this feature further up in the document, for the sake of consistency with the rest of the features?

aleksijuvani replied:
Done! I also noticed that this should probably be under the system information section rather than under windowing, so I've moved it there.

Osspial commented on src/platform_impl/windows/monitor.rs:
```diff
@@ -170,4 +175,44 @@ impl MonitorHandle {
+                    i,
+                    &mut mode,
+                    0,
+                ) == 0
```
Multi-line `if` clauses harm code readability. Could you pull this into a separate variable?

aleksijuvani replied:
I've moved `device_name` into a separate variable so this fits into a single line.

felixrabe commented:
I just noticed on my old macOS 10.9 MBAir that the `core-video-sys` crate brings in `metal` as a dependency, which means I can't compile at all.

Is there some way to do this without depending on Metal?

aleksijuvani commented:
We don't actually need Metal for this at all, but core-video-sys happens to pull that in for other reasons. I've opened LuoZijun/rust-core-video-sys#3 to make the Metal dependency optional. I will follow this up with a pull request to winit that then disables the "metal" feature for the core-video-sys dependency.
